File: core/ui/src/types/deviation.ts
import type { PageResponse } from "./grn";

export type DeviationSeverity = "MINOR" | "MAJOR" | "CRITICAL";
export type DeviationStatus = "OPEN" | "UNDER_INVESTIGATION" | "CAPA_IN_PROGRESS" | "CLOSED" | "CANCELLED";
export type DeviationSourceModule = "MANUAL" | "GRN" | "SAMPLING" | "INVENTORY" | "WAREHOUSE" | "BATCH";
export type DeviationType = "MATERIAL" | "PROCESS" | "DOCUMENTATION" | "EQUIPMENT" | "FACILITY" | "SAFETY" | "OTHER";

export type Deviation = {
  id: string;
  deviationNumber: string;
  title: string;
  description: string;
  deviationType: DeviationType;
  severity: DeviationSeverity;
  status: DeviationStatus;
  sourceModule: DeviationSourceModule;
  sourceEntityId: string | null;
  sourceReference: string | null;
  department: string | null;
  detectedBy: string;
  detectedAt: string;
  immediateAction: string | null;
  investigationSummary: string | null;
  rootCause: string | null;
  impactAssessment: string | null;
  closureSummary: string | null;
  closedBy: string | null;
  closedAt: string | null;
  closureESignatureId: string | null;
  isActive: boolean;
  createdBy: string;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
};

export type CreateDeviationRequest = {
  title: string;
  description: string;
  deviationType: DeviationType;
  severity: DeviationSeverity;
  sourceModule: DeviationSourceModule;
  sourceEntityId?: string;
  sourceReference?: string;
  department?: string;
  detectedAt?: string;
  immediateAction?: string;
};

export type UpdateDeviationRequest = {
  title: string;
  description: string;
  deviationType: DeviationType;
  severity: DeviationSeverity;
  department?: string;
  immediateAction?: string;
  investigationSummary?: string;
  rootCause?: string;
  impactAssessment?: string;
};

export type DeviationStatusUpdateRequest = {
  status: DeviationStatus;
  reason?: string;
  closureSummary?: string;
  username?: string;
  password?: string;
  meaning?: string;
};

export type DeviationSummary = {
  countsByStatus: Record<DeviationStatus, number>;
  countsBySeverity: Record<DeviationSeverity, number>;
};

export type DeviationPage = PageResponse<Deviation>;

File: core/src/main/java/com/batchsphere/core/qms/deviation/entity/Deviation.java
package com.batchsphere.core.qms.deviation.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "qms_deviation")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Deviation {

    @Id
    @Column(nullable = false, updatable = false)
    private UUID id;

    @Column(name = "deviation_number", nullable = false, unique = true, length = 100)
    private String deviationNumber;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "deviation_type", nullable = false, length = 40)
    private DeviationType deviationType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private DeviationSeverity severity;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 40)
    private DeviationStatus status;

    @Enumerated(EnumType.STRING)
    @Column(name = "source_module", nullable = false, length = 40)
    private DeviationSourceModule sourceModule;

    @Column(name = "source_entity_id")
    private UUID sourceEntityId;

    @Column(name = "source_reference", length = 120)
    private String sourceReference;

    @Column(length = 100)
    private String department;

    @Column(name = "detected_by", nullable = false, length = 100)
    private String detectedBy;

    @Column(name = "detected_at", nullable = false)
    private LocalDateTime detectedAt;

    @Column(name = "immediate_action", columnDefinition = "TEXT")
    private String immediateAction;

    @Column(name = "investigation_summary", columnDefinition = "TEXT")
    private String investigationSummary;

    @Column(name = "root_cause", columnDefinition = "TEXT")
    private String rootCause;

    @Column(name = "impact_assessment", columnDefinition = "TEXT")
    private String impactAssessment;

    @Column(name = "closure_summary", columnDefinition = "TEXT")
    private String closureSummary;

    @Column(name = "closed_by", length = 100)
    private String closedBy;

    @Column(name = "closed_at")
    private LocalDateTime closedAt;

    @Column(name = "closure_esignature_id")
    private UUID closureESignatureId;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @Column(name = "created_by", nullable = false, length = 100)
    private String createdBy;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}
